daemon/daemon.go

```diff
@@ -243,19 +243,19 @@ func (d *Daemon) RemoveNetworkPolicy(e *endpoint.Endpoint) {
 // start building.  The returned function must then be called to
 // release the "build permit" when the most resource intensive parts
 // of the build are done. The returned function is idempotent, so it
-// may be called more than once. Returns nil if the caller should NOT
+// may be called more than once. Returns a nil function if the caller should NOT
 // start building the endpoint. This may happen due to a build being
 // queued for the endpoint already, or due to the wait for the build
 // permit being canceled. The latter case happens when the endpoint is
-// being deleted.
-func (d *Daemon) QueueEndpointBuild(epID uint64) func() {
+// being deleted. Returns an error if the build permit could not be acquired.
+func (d *Daemon) QueueEndpointBuild(ctx context.Context, epID uint64) (func(), error) {
 	d.uniqueIDMU.Lock()
 	// Skip new build requests if the endpoint is already in the queue
 	// waiting. In this case the queued build will pick up any changes
 	// made so far, so there is no need to queue another build now.
 	if _, queued := d.uniqueID[epID]; queued {
 		d.uniqueIDMU.Unlock()
-		return nil
+		return nil, nil
 	}
 	// Store a cancel function to the 'uniqueID' map so that we can
 	// cancel the wait when the endpoint is being deleted.
```

Comment:
Does the line below this mess things up? ctx variable is overwritten, isn't it? and then the cancel from the ctx parameter won't ever be seen. In particular, I'm not sure that the `ctx.Err()` check would ever trigger then.

Reply:
Good catch, fixing.

Reply:
Fixed.